src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,210 @@
+    const formElements = Array.from(formNodes)
+
+    formElements.forEach((elem: any) => {
+      const inputId = elem.attributes['data-percy-element-id'].value
+      const selector = `[data-percy-element-id="${inputId}"]`
+      const cloneEl = clonedDOM.querySelector(selector) as HTMLInputElement
+
+      switch (elem.type) {
+        case 'checkbox':
+        case 'radio':
+          if (elem.checked && !elem.hasAttribute('checked')) {
+            cloneEl!.setAttribute('checked', '')
+          }
+          break
+        case 'textarea':
+          // setting text or value does not work but innerText does
+          if (elem.innerText !== elem.value) {
+            cloneEl!.innerText = elem.value
+          }
+        default:
+          if (!elem.getAttribute('value')) {
+            cloneEl!.setAttribute('value', elem.value)
+          }
+      }
+    })
+
+    return clonedDOM
+  }
+
+  /**
+   * Capture in-memory styles & serialize those styles into the cloned DOM.
+   *
+   * Without this, applications that use packages like `styled-components` will be
+   * missing styles and appear broken. The CSSOM provides a way to create CSS
+   * that only lives in memory (not an asset or in the DOM).
+   *
+   */
+  private serializeCssOm(documentClone: HTMLDocument) {
+    const styleSheets = Array.from(this.originalDOM.styleSheets)
+
+    styleSheets.forEach((styleSheet: any) => {
+      // Make sure it has a rulesheet, does NOT have a href (no external stylesheets),
+      // and isn't already in the DOM.
+      function isCSSOM() {
+        const hasHref = styleSheet.href
+        const ownerNode = styleSheet.ownerNode as HTMLElement
+        const hasStyleInDom =
+          ownerNode.innerText && ownerNode.innerText.length > 0
+
+        return !hasHref && !hasStyleInDom && styleSheet.cssRules
+      }
+
+      if (isCSSOM()) {
+        const $style = documentClone.createElement('style')
+        const cssRules = Array.from(styleSheet.cssRules)
+        const serializedStyles = cssRules.reduce((prev: string, cssRule: any) => {
+          return prev + cssRule.cssText
+        }, '')
+
+        // Append the serialized styles to the cloned document
+        $style.type = 'text/css'
+        $style.setAttribute('data-percy-cssom-serialized', 'true')
+        $style.innerHTML = serializedStyles
+        // TODO, it'd be better if we appended it right after the ownerNode in the clone
+        documentClone.head.appendChild($style)
+      }
+    })
+
+    return documentClone
+  }
+
+  /**
+   * A single place to mutate the origial DOM. This should be the last resort!
+   * This will change the customers DOM and have a possible impact on the
+   * customers application.
+   *
+   */
+  private mutateOriginalDOM() {
+    function createUID($el: Element) {
+      const ID =
+        '_' +
+        Math.random()
+          .toString(36)
+          .substr(2, 9)
+
+      $el.setAttribute('data-percy-element-id', ID)
```

Comment:
I guess it's safe to set these to new values even if they already existed?  i.e. a second snapshot on the same dom will replace the ids added by the first snapshot..

Reply:
We could probably save some work for ourselves by checking if it exists first. It's not too expensive to do it each time but I think it's worth checking first